File: rust/src/day02.rs
use lazy_static::lazy_static;
use regex::Regex;

static INPUT: &str = include_str!("../../inputs/day02.txt");

lazy_static! {
    static ref POLICY_REGEX: Regex = Regex::new(r"^(\d+)-(\d+) (.): (.+)$").unwrap();
}

type Policy<'a> = (usize, usize, char, &'a str);

pub fn run() {
    let policies = transform_input(INPUT);

    println!("Part 1: {}", part1(&policies));
    println!("Part 2: {}", part2(&policies));
}

fn transform_input(input: &str) -> Vec<Policy> {
    input
        .lines()
        .filter_map(|x| {
            let caps = POLICY_REGEX.captures(x)?;

            Some((
                caps.get(1)?.as_str().parse().unwrap(),
                caps.get(2)?.as_str().parse().unwrap(),
                caps.get(3)?.as_str().chars().next()?,
                caps.get(4)?.as_str(),
            ))
        })
        .collect()
}

fn part1(policies: &Vec<Policy>) -> usize {
    policies
        .iter()
        .filter(|&(low, high, pol_char, pass)| {
            let char_count = pass.chars().filter(|c| c == pol_char).count();
            &char_count >= low && &char_count <= high
        })
        .count()
}

fn part2(policies: &Vec<Policy>) -> usize {
    policies
        .iter()
        .filter(|&(low, high, pol_char, pass)| {
            let mut chars = pass.chars();
            let lchar = chars.nth(low - 1).unwrap();
            let hchar = chars.nth(high - low - 1).unwrap();

            (pol_char == &lchar) ^ (pol_char == &hchar)
        })
        .count()
}

File: haskell/src/Day02.hs
module Day02 where

import qualified Data.Either                as E
import           Data.Void                        (Void)
import qualified Text.Megaparsec            as P
import qualified Text.Megaparsec.Char       as P
import qualified Text.Megaparsec.Char.Lexer as L

type Parser = P.Parsec Void String

data Policy = Policy Int Int Char String
              deriving (Show)

run :: String -> IO ()
run input =
  do let policies = E.rights $ P.parse parser "" <$> lines input

     putStrLn $ "Part 1: " ++ show (part1 policies)
     putStrLn $ "Part 2: " ++ show (part2 policies)

parser :: Parser Policy
parser = Policy
         <$> L.decimal
         <*> (P.char '-' *> L.decimal)
         <*> (P.space *> P.lowerChar)
         <*> (P.chunk ": " *> P.takeRest)

part1 :: [Policy] -> Int
part1 = length . filter guard
  where
    guard :: Policy -> Bool
    guard (Policy low high char password) =
      let n = count char password
       in low <= n && n <= high

count :: Char -> String -> Int
count c = length . filter (==c)

part2 :: [Policy] -> Int
part2 = length . filter guard
  where
    guard :: Policy -> Bool
    guard (Policy low high char pass) =
      (pass !! (low - 1) == char) /= (pass !! (high - 1) == char)
